CoderiseIO: Update Documentation

adelanre opened the pull request:
- Fix typo in the /etc/rsyslog path
- Add comment in item #1 for command to run as sudo user.
- Add comment to `-p` to the mkdir command.

DavidAntaramian commented on kubernetes/configuration/index.md:
````diff
@@ -0,0 +1,30 @@
+
+    ```toml
+[kubernetes.exclude]
+namespaces = “kube-system”
````
Your system is using smart quotes; please convert these to straight quotes.

DavidAntaramian commented on kubernetes/configuration/index.md:
````diff
@@ -0,0 +1,30 @@
+    ```toml
+[kubernetes.exclude]
+namespaces = “kube-system”
+pods = “timber-agent”
````
Your system is using smart quotes; please convert these to straight quotes.

DavidAntaramian commented on kubernetes/index.md:
```diff
@@ -0,0 +1,12 @@
+
+With Timber's Kubernetes integration you can seamlessly deliver logs from your [Docker] Containers running on your
+[Kubernetes] Cluster to your Timber account. Delivery is achieved by using our
+[light weight highly efficient binary agent](https://timber.io/docs/platforms/other/agent).
```
This is a bit wordy. I would suggest:

> With Timber's Kubernetes integration, you can seamlessly deliver logs from pods in your [Kubernetes] Cluster to your Timber account.

The delivery method is an implementation detail that the overview doesn't need to cover.

You _may_ want to add details about the information that's included when you use the integration.

DavidAntaramian commented on kubernetes/installation/index.md:
````diff
@@ -0,0 +1,54 @@
+1. Create Kubernetes Secret with Timber API Key
+
+  ```bash
+  kubectl create secret generic timber --from-literal=timber-api-key=API_KEY
````
I believe that you need to replace `API_KEY` with `{{timber_api_key}}` for the server-side interpolation to automatically replace it with the API key, but @binarylogic can speak better to that.

binarylogic replied:
That's correct, for now I would do that. Going forward we'll have a better way of accomplishing this since we'll more advanced capabilities with Gitdocs. I imagine something similar to Stripe's docs: http://bens.files.timber.io/0F2X2i452i2U

DavidAntaramian commented on kubernetes/installation/index.md:
````diff
@@ -0,0 +1,54 @@
+  _2a. For Kubernetes Cluster versions >= 1.7_
+
+  ```bash
+  kubectl apply -f https://gist.githubusercontent.com/carkmorwin/1bffc24918a20802c8fce51fde865d3f/raw/c14d4093a8123066b299a86fb645ec5ab00c176e/timber-agent-daemonset.yaml
````
I would recommend moving this file into the Agent repository at `support/scripts/kubernetes/timber-agent-daemonset.yaml` and then using the appropriate raw-format link for a tagged version of the file.

DavidAntaramian commented on kubernetes/installation/index.md:
````diff
@@ -0,0 +1,54 @@
+  _2b. For Kubernetes Cluster versions <= 1.6_
+
+  ```bash
+  kubectl apply -f https://gist.githubusercontent.com/carkmorwin/24e08aa51c23999b177e42cbbbf1f7dd/raw/b402a518f3664c07325b5991cae47a97fd5181d4/timber-agent-daemonset-legacy.yaml
````
I would recommend moving this file into the Agent repository at `support/scripts/kubernetes/timber-agent-daemonset-legacy.yaml` and then using the appropriate raw-format link for a tagged version of the file.

DavidAntaramian commented on kubernetes/installation/index.md:
````diff
@@ -0,0 +1,54 @@
+Create RBAC resources
+
+```bash
+kubectl apply -f https://gist.githubusercontent.com/carkmorwin/e624cc9985011b818b0c1f32ba320ec6/raw/61098a074accf33ecce7db37ad84f51787a37a3f/timber-agent-daemonset-with-rbac.yaml
````
I would recommend moving this file into the Agent repository at `support/scripts/kubernetes/timber-agent-daemonset-rbac.yaml` and then using the appropriate raw-format link for a tagged version of the file.

DavidAntaramian commented on kubernetes/installation/index.md:
```diff
@@ -0,0 +1,54 @@
+
+If you are managing RBAC outside of this install, the [Timber Agent ClusterRole] manifest can be found here.
+
+https://gist.githubusercontent.com/carkmorwin/3fd533213f018776f7ba5f0ffdea5d15/raw/4357aab7e3cf743ecf06a3ee21f99b79947a0ae4/timber-agent-clusterrole.yaml
```
I would recommend moving this file into the Agent repository at `support/scripts/kubernetes/timber-agent-clusterrole.yaml` and then using the appropriate raw-format link for a tagged version of the file.

DavidAntaramian commented on kubernetes/troubleshooting/index.md:
````diff
@@ -0,0 +1,36 @@
+
+You can verify pods are running via `kubectl get pods -l name=timber-agent`.
+
+```bash
````
This output is just text, not `bash`, so you should drop the language identity here

DavidAntaramian commented on kubernetes/troubleshooting/index.md:
````diff
@@ -0,0 +1,36 @@
+
+If not running or initializing, you can view more details about failures with kubectl.
+
+```
````
This is `bash`, so you should add the language identity here.

DavidAntaramian commented on kubernetes/troubleshooting/index.md:
```diff
@@ -0,0 +1,36 @@
+
+- __Pods fail to start__
+
+You can verify pods are running via `kubectl get pods -l name=timber-agent`.
```
Place this command in a separate code-fence rather than inline to match the layout of the rest of the document.

DavidAntaramian commented on kubernetes/troubleshooting/index.md:
````diff
@@ -0,0 +1,36 @@
+kubectl describe pods -l name=timber-agent
+```
+
+Debugging is outside the scope of these docs, but hopefully you should enough information to navigate the Kubernetes docs with.
````
I would rephrase this:

> This information should help you understand why the integration is not being started in your cluster. Errors with the cluster configuration are outside the scope of this documentation, and you should refer to the official Kubernetes documentation and community for further assistance.

DavidAntaramian commented on kubernetes/troubleshooting/index.md:
````diff
@@ -0,0 +1,36 @@
+kubectl logs -l name=timber-agent -c timber-agent
+```
+
+Check this output for warning or error messages. Additionally, your file may have matched an exclusion filter. Info logs are printed with `File logs will not be forwarded due to matching an exclusion filter` with which file was excluded and what filter matched.
````
Replace:

> Additionally, your file may have matched an exclusion filter. Info logs are printed with `File logs will not be forwarded due to matching an exclusion filter` with which file was excluded and what filter matched.

With something like:

> Additionally, your file may have matched an exclusion filter. If this is the case, there will be a log entry similar to the following to inform you: `File app_container-b will not be forwarded due to matching an exclusion filter`.

DavidAntaramian commented on kubernetes/troubleshooting/index.md:
```diff
@@ -0,0 +1,36 @@
+
+- __Failed to read metadata__
+
+When the Kubernetes API is unreachable, retries will be performed until it is available again. In the case this lasts a significantly long time is is unavailable from the start, there may be an issue with the kubectl proxy setup or the node itself. Viewing the kubectl proxy logs should indicate the error, or if running, verify the proxy config is correct for the timber agent container.
```
I would replace the above with this:

> The integration has a soft dependency on the Kubernetes API to gather facts about the containers logs are collected from. If the Kubernetes API is unavailable when a request is made, the integration will retry the request (using an exponential backoff) until the service is available again.
>
> The integration will continue to forward logs to the Timber service even when the Kubernetes API is unavailable. These logs, though, will not have additional metadata about the cluster, and you will see "Failed to read metadata" in the integration logs.
>
> If the Kubernetes API is unavailable for a significantly long period of time or is unavailable from the start, this usually means there is an issue with the kubectl proxy setup or the node itself. The kubectl proxy logs should indicate if an error has occurred. If the logs do not show an error, verify that the proxy configuration is correct for the Timber Agent container.

That being said, two open issues remain:

  - [ ] How long is "a significantly long time"? That could mean 15 minutes or 3 years.
  - [ ] How does the user "verify the proxy config is correct for the timber agent container"?

DavidAntaramian commented on kubernetes/configuration/index.md:
```diff
@@ -0,0 +1,30 @@
+
+# Configuration
+
+Timber’s Kubernetes integration options are supplemental to the Timber Agent configuration, under the table `[kubernetes]`.
```
I wouldn't use the term `table` here, since that would expect the user to understand TOML terminology. I would just say "under the `[kubernetes]` configuration section".

carkmorwin commented:
@DavidAntaramian just wanted to say all of your comments are __extremely__ helpful and very much appreciated.

DavidAntaramian commented:
@jsonmaur and @zsherman could you please make changes in a separate branch? You could base it off this one, but I'd like to prevent Mark from having to handle merge conflicts on his end when you push changes, and your changes are also not relevant to the PR topic.

jsonmaur commented:
Woops, yes sorry.

binarylogic commented on kubernetes/configuration/index.md:
```diff
@@ -0,0 +1,30 @@
+
+Configuration Options
+
+- `exclude` __(table)__
```
Same thing here regarding the "table" term.

binarylogic commented on kubernetes/configuration/index.md:
```diff
@@ -0,0 +1,30 @@
+---
+Title: Configuration
```
This document is a great candidate for the multiple repo composition feature. Documentation that relates directly to the agent and it's configuration should live within the agent repository. For example, I imagine we'd have a file `/docs/configuration/kubernetes.md` within the agent repository itself.

@zsherman @jsonmaur is feature currently available? Is there an example of how to specify a markdown file that lives in another repository?

DavidAntaramian replied:
On that basis, I almost feel like this entire change should be moved to the agent repo, so you'd have `docs/configuration.md` (which has general configuration) and `docs/kubernetes/installation.md` and `docs/kubernetes/configuration.md` which holds the Kubernetes specific configuration but can link back to the main configuration document.

That's also more in line with the idea of keeping the documentation as close to the code as possible. In other words, when the method for installing Kubernetes changes etc, the documentation can be changed in the same PR.

jsonmaur replied:
For a single file, you can put a `source` key in the front matter with a static URL. It will make a simple GET request and inject the response into the page. e.g.
```
---
source: https://raw.githubusercontent.com/timberio/agent/master/README.md
source_inject: after
// or
source_inject: before
// or
source_inject: replace (the default)
---
```
We also have the git repo sources in place, but still working out the kinks. This will all be documented soon.

binarylogic commented on kubernetes/installation/index.md:
```diff
@@ -0,0 +1,54 @@
+
+RBAC Support (1.6+ required)
+
+_It is assumed you have have enabled and verified RBAC._
```
have have

binarylogic commented on kubernetes/index.md:
```diff
@@ -0,0 +1,12 @@
+# Kubernetes
+
+With Timber's Kubernetes integration, you can seamlessly deliver logs from pods in your [Kubernetes] Cluster to your
+Timber account. These logs are shipped with supplementary Kubernetes metadata that assists in identifying and grouping
```
We really need to formalize how we introduce users to Timber metadata within each platform's documentation. For example, this should probably link to a general metadata document as well as the specific Kubernete's fields that this integration offers.

You don't need to make this change within this PR, but I wanted to mention it.

DavidAntaramian replied:
I think a higher level concept document would be good.

jsonmaur commented:
@binarylogic GitDocs isn't currently set up to detect an index file that way. It will look for an `index.md` in the folder when building the navigation structure, which is similar to how static sites work with an `index.html`, and will output a warning if done the way you mentioned.

This should probably be thought out better, though. [I opened an issue for it](https://github.com/timberio/gitdocs/issues/96), let's move the discussion over there.

jsonmaur commented:
To clarify, this only applies to the index file of the folder, which is optional. In this case, you don't need to have an index for each folder, you could just do
```
docs/kubernetes/index.md
docs/kubernetes/configuration.md
```

binarylogic commented:
👍 . My concern was that some platforms aren't necessarily tied to the agent. For example, where do platforms using syslog go? The agent has nothing to do with those.

carkmorwin commented:
👍 moving docs to agent repo.

carkmorwin commented:
@jsonmaur @zsherman I left your commits here. Pull them down or move them somewhere and close this PR. Thanks.

adelanre committed:
CoderiseIO: Update Documentation

- Fix typo in the /etc/rsyslog path
- Add comment in item #1 for command to run as sudo user.

adelanre committed:
Update syslog.md

Add check rsyslog status

adelanre committed:
Update syslog.md

Update comment on rsyslog-gnutls for CentOS

adelanre committed:
Update syslog.md

Add yum install command

adelanre committed:
Update syslog.md

Fix typo in rsyslog status command
